### PR TITLE
Controller behavior interface

### PR DESCRIPTION
**Pull request by derekisbusy**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13473

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,7 @@ Yii Framework 2 Change Log
 
 - Bug #13501: Fixed `yii\rbac\DbManager::getRule()` and `yii\rbac\DbManager::getRules()` to properly handle resource data came from Rule table when using PostgreSQL (StalkAlex)
 - Bug #13508: Fixed duplicate attachment of behavior BC break (cebe)
+- Enh #13473: Add ControllerBehaviorInterface that allows behaviors who implement the interface to add actions to their owner. (derekisbusy)
```
Comment (arogachev) [inline framework/CHANGELOG.md]: "Who" is never used to refer to things. See for example [this article](http://blog.oxforddictionaries.com/2012/09/that-who-which/) for more details.

File: framework/base/ControllerBehaviorInterface.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+namespace yii\base;
+/**
+ * ControllerBehaviorInterface is the class that should be implemented by behaviors who want to add actions to the controller.
+ *
+ * @author Qiang Xue <qiang.xue@gmail.com>
```
Comment (arogachev) [inline framework/base/ControllerBehaviorInterface.php]: Write your name and email instead or just omit that.

File: framework/base/ControllerBehaviorInterface.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+namespace yii\base;
+/**
+ * ControllerBehaviorInterface is the class that should be implemented by behaviors who want to add actions to the controller.
```
Comment (arogachev) [inline framework/base/ControllerBehaviorInterface.php]: Again, wrong usage of "who".

File: framework/CHANGELOG.md [hunk 4]
```diff
@@ -6,6 +6,7 @@ Yii Framework 2 Change Log
 
 - Bug #13501: Fixed `yii\rbac\DbManager::getRule()` and `yii\rbac\DbManager::getRules()` to properly handle resource data came from Rule table when using PostgreSQL (StalkAlex)
 - Bug #13508: Fixed duplicate attachment of behavior BC break (cebe)
+- Enh #13473: Add ControllerBehaviorInterface that allows behaviors which implement the interface to add actions to their owner. (derekisbusy)
```
Comment (samdark) [inline framework/CHANGELOG.md]: Should be moved to 2.0.12...

Comment (samdark) [inline framework/CHANGELOG.md]: Last `.` before nickname should be removed.

File: framework/base/Controller.php [hunk 5]
```diff
@@ -221,6 +221,13 @@ public function createAction($id)
         }
 
         $actionMap = $this->actions();
+        $this->ensureBehaviors();
+        foreach ($this->getBehaviors() as $behavior) {
+            if ($behavior instanceof ControllerBehaviorInterface) {
+                $actionMap = array_replace_recursive($actionMap, $behavior->actions());
```
Comment (samdark) [inline framework/base/Controller.php]: Not sure if behavior should replace already defined actions.

File: framework/base/ControllerBehaviorInterface.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+namespace yii\base;
+/**
+ * ControllerBehaviorInterface is the class that should be implemented by behaviors that want to add actions to the controller.
+ *
+ * @author Derek Gifford <derekisbusy@gmail.com>
+ * @since 2.0.11.2
```
Comment (samdark) [inline framework/base/ControllerBehaviorInterface.php]: 2.0.12

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/base/Controller.php (hunk 5 above)
Comment (derekisbusy) [inline framework/base/Controller.php]: Could change it to use `ArrayHelper::merge()` and add a new helper `yii\helpers\ArrayValueIfKeyNotExists` that would only add the key/value pair if the key doesn't exist in the other array.

```php
$array1 = [
        'error' => 'yii\web\ErrorAction',
        'view' =>  'yii\web\ViewAction',
 ];
$array2 = [
        'error' => 'yii\web\ErrorAction',
        'view' => new \yii\helpers\ArrayValueIfKeyNotExists(  'my\custom\ViewAction'),
        'custom' => new \yii\helpers\ArrayValueIfKeyNotExists(  'my\custom\CustomAction'),
 ];

$result = ArrayHelper::merge($array1, $array2);

// result will be...
[
        'error' => 'yii\web\ErrorAction',
        'view' =>  'yii\web\ViewAction',
        'custom' => 'my\custom\CustomAction',
 ];
```

This would provide the most flexibility as it would allow the developer to decide whether or not the values should be replaced.

Comment (derekisbusy) [inline framework/base/Controller.php]: Could also add a second parameter to control what happens if the key already exists to make debugging less ambiguous:

```php
$array1 = [
        'error' => 'yii\web\ErrorAction',
        'view' =>  'yii\web\ViewAction',
        'custom' => 'my\custom\CustomAction',
 ];
$array2 = [
        'error' => 'yii\web\ErrorAction',
        'view' => new \yii\helpers\ArrayValueIfKeyNotExists(  'my\custom\ViewAction' , function() { 
                               throw new Exception('View action already defined.'); 
        }),
        'custom' => new \yii\helpers\ArrayValueIfKeyNotExists(  'my\custom\CustomAction'),
 ];

// would throw an exception
$result = ArrayHelper::merge($array1, $array2);

```

**Comment by rob006** [comment]
This makes actions really magically and hard to follow. Right now if you want check what actions have specified controller, you search methods prefixed by `action` and `action()` method. After this PR you need check additionally each behavior. And if you use vendor behavior, each update could implicitly add new actions to controller.

**Comment by samdark** [comment]
@rob006 yes, it becomes as implicit as behaviors are.

**Comment by rob006** [comment]
The same could be achieved with

```php
    public function actions()
    {
        return array_merge(
            $this->getBehavior('myBehavior1')->actions(),
            $this->getBehavior('myBehavior2')->actions(),
            $this->getBehavior('myBehavior3')->actions(),
            [...],
        );
    }
```

in controller class, but with more control over actions and less magic.

**Comment by derekisbusy** [comment]
@rob006 The only advantage would be that you can add actions to a controller without updating the controller, for example controllers in vendor extensions. But this can also be achieved by extending the controller using the controller map. So I guess the proper way to do this would be to extend the controller and add the behavior and actions.
